File: python/minilisp/eval.py
from minilisp import data
from minilisp import forms


def evaluate(env: data.Environment, expr: data.Value) -> data.Value:
    # print('DEBUG: evaluate: %s' % expr)
    if isinstance(expr, (data.Null, data.Bool, data.Int)):
        return expr
    if isinstance(expr, data.Symbol):
        return env.lookup(expr.name).value
    if isinstance(expr, data.Pair):
        if isinstance(expr.car, data.Symbol):
            name = expr.car.name
            form_func = forms.ALL_MAP.get(name)
            if form_func:
                args = data.to_native_list(expr.cdr)
                return form_func(env, args)
        func = evaluate(env, expr.car)
        args = [evaluate(env, value) for value in data.to_native_list(expr.cdr)]
        return func.apply(args)
    raise AssertionError('Can not evaluate: %s' % expr)

File: javascript/minilisp/eval.js
const data = require('./data.js');

function evaluate(env, expr) {
  if (expr instanceof data.Undef || expr instanceof data.Bool || expr instanceof data.Int) {
    return expr;
  }
  if (expr instanceof data.Symbol) {
    return env.lookup(expr.name).value
  }
  if (expr instanceof data.Pair) {
    const rawArgs = data.valueToArray(expr.cdr);
    if (!(expr.car instanceof data.Symbol)) {
      throw new Error('no form implemented');
    }
    const func = evaluate(env, expr.car);
    if (!(func instanceof data.Func)) {
      throw new Error('can not call non-function value');
    }
    const args = rawArgs.map((rawArg) => evaluate(env, rawArg));
    return func.call(args);
  }
  throw new Error('not evaluatable value');
}

module.exports = {
  evaluate,
};
